Release lifecycle marking (#225)
signalfx/gdi-specification: specification/versioning.md

  MUST be provided as the latest `PATCH` version for the latest `MINOR` version
  of the latest `MAJOR` and SHOULD NOT be provided for previous `PATCH` or
  `MINOR` releases.

## Release life cycle management

Each repository `MUST` use [GitHub releases mechanism](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases) to create snapshot of components
and features delivered to the users. If release is also distributed via other 
distribution channels (e.g. CDN, Registries, AppStores) it `MUST` be described 
according to following rules by any means available in the given software 
distribution solution. 
 
Each release `MUST` clearly state the life cycle 

> status and described as: Active, Latest, Deprecated, End of Support. 

Author: It is possible to use the same list of labels/statuses on levels however the meaning might be little different depending on the context also the application might be different so to keep it clear we would have to map each status with a description on what it actually means in each context (release, component, spec etc) The final list would look like this:
(None) 
Experimental
Stable (same as active)
Deprecated 
End of Support 
Removed 

Additionally it should be possible to annotate Deprecated and End of Support with a date

> status and described as: Active, Latest, Deprecated, End of Support. 

Reviewer: Resolving per https://github.com/signalfx/gdi-specification/pull/225#discussion_r1153220625

> status and described as: Active, Latest, Deprecated, End of Support. 

Author: The main issue is the context of in which "Stable" and "Experimental" is used - it work well for functional maturity it works less for lifecycle e.g. "this release has stable support", "this release has experimental support" or is "supported as experimental" thus was the proposal to keep a parallel set of labels. As stated in other comment we can make use one set but we need to explain what they mean depending on which context we use them in

> status and described as: Active, Latest, Deprecated, End of Support. 

Reviewer: I like this idea 👍

> At minimum the required information should be included in [CHAANGELOG.md](https://github.com/signalfx/gdi-specification/blob/main/specification/templates/CHANGELOG.md)

Reviewer: ```suggestion
At minimum the required information should be included in 
[CHANGELOG.md](https://github.com/signalfx/gdi-specification/blob/main/specification/templates/CHANGELOG.md)
```

and equivalent mechanism provided by system used for distributing software. 

If new version is released previously released versions `MUST` be evaluated 
according to stability guarantees and marked as:
- Latest release `SHOULD` be marked as **Latest** indicating that this is

>   version recommended to new and existing users

Reviewer: Should this not be only applied for "stable" releases. E.g. beta/rc releases are usually not marked as latest.

Author: Good example - we should keep the latest at the stable level as we do not want to promote latest on experimental releases. The new release  release should be marked as experimental or experimental, supported 
Btw no I think about it - instead of "Active" we should probably use "Supported"

Author: So actually instead of 
-	*Latest*
-	*Active*
-	*Deprecated:* End of Support expected on 2024-04-01
-	*End of Support:* Support has ended on 2024-04-01

We could have 
-       "Supported"
-	*Deprecated:* End of Support expected on 2024-04-01
-	*End of Support:* Support has ended on 2024-04-01

Author: @pellared change introduced and committed to PR. Please take a look

- If applicable latest release of previous `MAJOR` line that is still under 
  active development `MUST` be marked as **Active** indicating that previous 
  `MAJOR` line is under active development and users not ready to adopt 
  latest `MAJOR` version may use previous versions.
- Release which was superseded by newer version, thus has entered deprecation 
  period `MUST` be marked as **Deprecated** and `MUST` indicated planned 
  end of support date. This indicates that users should plan for upgrade to 

>   latest or active version respectively before date given.

Author: Minor and Major:
1. Major version can be used and live for more than 12M
2. In some cases Major releases may be rare

- Release which deprecation period has ended `MUST` be marked as 
  **End of support** and `MUST` indicate date on which the release was marked 
  as end of support. This indicates that support for given release is no longer 

>   provided and the users should immediately upgrade to newer version.

Reviewer: Can we remove "releases" which are Out of Support? E.g. some package management systems offers retracting packages.

Author: We could once the EoS is reached we should probably have some short period after which we could bury a the release and l leave only tombstone behind or get rid of it all together

  MUST be provided as the latest `PATCH` version for the latest `MINOR` version
  of the latest `MAJOR` and SHOULD NOT be provided for previous `PATCH` or
  `MINOR` releases.

## Release life cycle management

Each repository `MUST` use [GitHub releases mechanism](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases)

> to create a versioned snapshot of components and features

Reviewer 2: ```suggestion
to publish versioned artifacts of components and features
```

I'd like to avoid "snapshot" here because it has a very specific meaning, at least in Java.

Reviewer: I updated my suggestion https://github.com/signalfx/gdi-specification/pull/225#discussion_r1159428244 to reflect your proposal

delivered to the users. If the release is also distributed via other
distribution channels (e.g. CDN, Registries, AppStores) it `MUST` state
its life cycle status according to following rules by any means available
in the given software distribution solution.

Each release `MUST` clearly state the life cycle

> status and described as: Supported, Deprecated, End of Support.

Reviewer: If we use "Obsolete" instead of "Deprecated" then we would have distinct status names for "Maturity" and "Releases". What do you think?

Reviewer 3: I think "Obsolete" is too negative and implies that the thing no longer works, whereas something Deprecated can still work while not being the favored choice.

At minimum the required information should be included in [CHAANGELOG.md](https://github.com/signalfx/gdi-specification/blob/main/specification/templates/CHANGELOG.md)
and equivalent mechanism provided by system used for distributing software.

If new version is released previously released versions `MUST` be evaluated
according to stability guarantees and marked as:

- Latest release `SHOULD` be marked as **Supported** indicating that this is
  version recommended to new and existing users
- If applicable latest release of previous `MAJOR` line that is still under
  active development `MUST` be marked as **Supported** indicating that previous
  `MAJOR` line is under active development and users not ready to adopt
  latest `MAJOR` version may use previous versions.
- Release which was superseded by newer version, thus has entered deprecation
  period `MUST` be marked as **Deprecated** and `MUST` indicated planned

>   end of support date. This indicates that users should plan for upgrade to

Reviewer: Do we have already a requirement how long it must be supported? If so can we define it? Otherwise it will be hard to adopt the specification.

What if the latest release fixes a security vulnerability? Shouldn't we immediately drop supportand require a version bump?

> Each repository `MUST` use [GitHub releases mechanism](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases)
> to create a versioned snapshot of components and features
> delivered to the users. If the release is also distributed via other
> distribution channels (e.g. CDN, Registries, AppStores) it `MUST` state
> its life cycle status according to following rules by any means available
> in the given software distribution solution.
> 
> Each release `MUST` clearly state the life cycle
> status and described as: Supported, Deprecated, End of Support.
> At minimum the required information should be included in [CHAANGELOG.md](https://github.com/signalfx/gdi-specification/blob/main/specification/templates/CHANGELOG.md)
> and equivalent mechanism provided by system used for distributing software.
> 
> If new version is released previously released versions `MUST` be evaluated
> according to stability guarantees and marked as:
> 
> - Latest release `SHOULD` be marked as **Supported** indicating that this is
>   version recommended to new and existing users
> - If applicable latest release of previous `MAJOR` line that is still under
>   active development `MUST` be marked as **Supported** indicating that previous
>   `MAJOR` line is under active development and users not ready to adopt
>   latest `MAJOR` version may use previous versions.
> - Release which was superseded by newer version, thus has entered deprecation
>   period `MUST` be marked as **Deprecated** and `MUST` indicated planned
>   end of support date. This indicates that users should plan for upgrade to
>   one of supported versions before date given.
> - Release which deprecation period has ended `MUST` be marked as
>   **End of support** and `MUST` indicate date on which the release was marked
>   as end of support. This indicates that support for given release is no longer
>   provided and the users should immediately upgrade to newer supported version.

Reviewer: I made some nit changes, remove some IMO unnecessary sentences or wording, I tried to restructure it a little to make it easier to read and validate when adopting the spec.

```suggestion
Each release MUST clearly state the life cycle
status and described as one of:
**Supported**, **Deprecated**, **End of Support**.
At minimum the required information is included in [CHANGELOG.md](templates/CHANGELOG.md)
and equivalent mechanism provided by system used for distributing software.

Each repository MUST use [GitHub releases mechanism](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases)
to create a versioned snapshot of components
delivered to the users.

The release the life cycle status MUST be included
in [CHANGELOG.md](templates/CHANGELOG.md).

When the component is also distributed via other channels
(e.g. CDN, Registries, AppStores) it MUST state
its life cycle status according to following rules by any means available
in the given software distribution solution.

When a new version is released,
all previously released versions MUST be evaluated
according to stability guarantees and marked as:

- If applicable latest release of previous `MAJOR` line that is still under
  active development MUST remain as **Supported** indicating that previous
  `MAJOR` line is under active development and users not ready to adopt
  the latest `MAJOR` version may use previous versions.

- Release which was superseded by newer version, thus has entered deprecation
  period `MUST` be marked as **Deprecated** and MUST indicated planned
  end of support date. This indicates that users should plan for upgrade to
  one of supported versions before date given.

- Release which deprecation period has ended MUST be marked as
  **End of support** and MUST indicate date on which the release was marked
  as end of support. This indicates that support for given release is no longer
  provided and the users should immediately upgrade to newer supported version.
```
